sklearn/feature_selection/_base.py

```diff
@@ -78,15 +79,22 @@ def transform(self, X):
         X_r : array of shape [n_samples, n_selected_features]
             The input samples with only the selected features.
         """
-        # note: we use _safe_tags instead of _get_tags because this is a
-        # public Mixin.
-        X = self._validate_data(
-            X,
-            dtype=None,
-            accept_sparse="csr",
-            force_all_finite=not _safe_tags(self, key="allow_nan"),
-            reset=False,
-        )
+        output_config_dense = _get_output_config("transform", estimator=self)["dense"]
+        if hasattr(X, "iloc") and output_config_dense == "pandas":
+            # Only check feature names and n_features when output is a dataframe
+            # This allow _transform to preserve `X`'s dtype
+            self._check_feature_names(X, reset=False)
+            self._check_n_features(X, reset=False)
+        else:
+            # note: we use _safe_tags instead of _get_tags because this is a
+            # public Mixin.
+            X = self._validate_data(
+                X,
+                dtype=None,
+                accept_sparse="csr",
+                force_all_finite=not _safe_tags(self, key="allow_nan"),
+                reset=False,
+            )
```

Comment:
can we move this logic somewhere which would be easy for third party developers to re-use? I'd be okay with something private around the `_check_feature_names`